### PR TITLE
sql-parser: Add new EXPLAIN code path.

### LINKED ISSUE #13294: Add new `EXPLAIN` code path to the SQL parser
1. Implement the grammar listed below in `parse_explain_new`.
2. In theory this should also include the change `sql-grammar.bnf`. However, this will break existing, so I suggest to punt this to #13298, revealing only the sub-grammar that is relevant to end users.

Since the output is (and will continue to be) very different from Postgres, and EXPLAIN is not part of the SQL standard, we take the liberty of continuing to deviate from the Postgres syntax for EXPLAIN after the refactor. 

```sql
-- new EXPLAIN PLAN syntax, introduced as part of #13139
-- result always contains exactly one row
-- output schema:
--   $plan_type: string not null
EXPLAIN
  ( RAW | DECORRELATED | OPTIMIZED | PHYSICAL )? format?
PLAN 
  with_config?
FOR
  (select_stmt | VIEW view_name)

-- syntax for traces, should be introduced as part of #13139
-- result contains one row for each applied transform (unique path)
-- and one additional row for the input plan
-- output schema:
--   plan: string not null
--   path: string not null "/phi/physical"
--   time: long not null
--   rule: string           -- maybe obsolete, as it can be the last element of the pass
EXPLAIN
  OPTIMIZER format?
TRACE
  with_config?
FOR
  (select_stmt | VIEW view_name)

format = TEXT | JSON | DOT | ...
with_config = WITH '(' ( attribute  )*[','] ')' 

-- examples: EXPLAIN PLAN
EXPLAIN RAW TEXT PLAN WITH ( monotonicity, types ) FOR ...
EXPLAIN OPTIMIZED DOT PLAN FOR ...
EXPLAIN PHYSICAL JSON PLAN WITH ( est_cost ) FOR ...

-- examples: EXPLAIN OPTIMIZER TRACE PLAN
EXPLAIN OPTIMIZER TEXT TRACE WITH ( monotonicity, types ) FOR ...
EXPLAIN OPTIMIZER DOT TRACE FOR ...
EXPLAIN OPTIMIZER JSON TRACE WITH ( est_cost ) FOR ...
```

### PR DESCRIPTION
**Pull request by wangandi**
TODO: we forgot when planning explain format consolidation what do about:
* EXPLAIN QUERY GRAPH
* EXPLAIN OPTIMIZED QUERY GRAPH
* EXPLAIN TIMESTAMP.

My opinion is that we should keep `EXPLAIN QUERY GRAPH` and `EXPLAIN OPTIMIZED QUERY GRAPH` in the new explain interface, but `EXPLAIN TIMESTAMP FOR ...` should be changed to `EXPLAIN ... PLAN WITH (timestamp) FOR ...` because I perceive timestamps as an attribute that can be appended onto a plan just like column types, but @mjibson is the authority on `EXPLAIN TIMESTAMP`.

### Motivation

  * This PR adds a known-desirable feature.

Closes #13294, which is part of #13137.

### Tips for reviewer

* I renamed our existing `ExplainStage` to `OldExplainStage` and created a separate `NewExplainStage`.
* I changed some `unimplemented!()` calls to be `Err(anyhow!("unimplemented interface")` calls to prevent `environmentd` from crashing if someone manages to stumble into the new explain interface.

### Testing

I added some positive tests for the new explain interface, but I didn't add negative tests. I think negative tests should be added after the old explain interface is removed.

### Release notes

No user-facing behavior changes.

### CODE REVIEW

File: src/sql-parser/src/ast/defs/statement.rs [hunk 1]
```diff
@@ -1803,20 +1803,38 @@ impl_display_t!(ExplainStatement);
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct ExplainStatementNew<T: AstInfo> {
+    pub stage: NewExplainStage,
+    pub format: ExplainFormat,
+    pub configs: Vec<Ident>,
```
Comment (aalexandrov) [inline src/sql-parser/src/ast/defs/statement.rs:1808]: nit: can we call this `config_flags`?

File: src/sql-parser/src/parser.rs [hunk 2]
```diff
@@ -4837,7 +4837,80 @@ impl<'a> Parser<'a> {
...
+        let configs = if self.parse_keyword(WITH) {
+            self.expect_token(&Token::LParen)?;
+            let configs = self.parse_comma_separated(Self::parse_identifier)?;
+            self.expect_token(&Token::RParen)?;
+            configs
+        } else {
+            vec![]
+        };
+
+        // TODO (#13137): When getting rid of the old explain syntax,
```
Comment (aalexandrov) [inline src/sql-parser/src/parser.rs:4894]: nit: can you change these TODO lines to reference #13299 as opposed to #13137 (this is the cleanup task for this epic).

File: src/sql-parser/tests/testdata/explain [hunk 3]
```diff
@@ -136,3 +136,26 @@ EXPLAIN TIMESTAMP FOR SELECT 1
 EXPLAIN TIMESTAMP FOR SELECT 1
 =>
 Explain(Old(ExplainStatementOld { stage: Timestamp, explainee: Query(Query { ctes: [], body: Select(Select { distinct: None, projection: [Expr { expr: Value(Number("1")), alias: None }], from: [], selection: None, group_by: [], having: None, options: [] }), order_by: [], limit: None, offset: None }), options: ExplainOptions { typed: false, timing: false } }))
+
+parse-statement
+EXPLAIN OPTIMIZED QUERY GRAPH TEXT PLAN WITH (monotonicity, types) FOR VIEW foo
```
Comment (aalexandrov) [inline src/sql-parser/tests/testdata/explain:141]: nit: in the original syntax, `QUERY GRAPH` is an alternative suffix to `PLAN`

```
EXPLAIN OPTIMIZED QUERY GRAPH
EXPLAIN OPTIMIZED PLAN
```

so we are not consistent with that syntax here. Maybe we can fix that?

File: src/sql-parser/src/parser.rs [hunk 4]
```diff
@@ -4837,7 +4837,80 @@ impl<'a> Parser<'a> {
...
+        // TODO (#13137): Make specifying the format optional upon getting rid
+        // of the old explain syntax.
+        let format = match self.parse_one_of_keywords(&[TEXT, JSON]) {
+            Some(TEXT) => ExplainFormat::Text,
+            Some(JSON) => ExplainFormat::Json,
+            None => return Err(ParserError::new(self.index, "expected a format")),
+            _ => unreachable!(),
+        };
+
+        if stage == NewExplainStage::Trace {
```
Comment (aalexandrov) [inline src/sql-parser/src/parser.rs:4876]: Ouch, I didn't realize that putting the format identifier before the noun would result in this code. The `parse_explain_old` nicely has `self.expect_keywords(&[...])?` in for each stage local to the if branch that is parsing this stage.

**Comment by wangandi** [comment]
> I now realize that making the `FORMAT` clause optional just before the noun (`QUERY GRAPH` / `PLAN` / `TRACE`) makes both the parser and the `*.bnf` diagrams substantially harder to write / comprehend.

I went for:
```
EXPLAIN OPTIMIZED QUERY GRAPH WITH (...) AS JSON ...
EXPLAIN OPTIMIZED PLAN AS TEXT ...
EXPLAIN OPTIMIZER TRACE WITH (...) AS TEXT ...
```
because @mjibson believes that `<thing to be explained>` should be first after the `EXPLAIN` verb.

**Comment by aalexandrov** [comment]
> > I now realize that making the `FORMAT` clause optional just before the noun (`QUERY GRAPH` / `PLAN` / `TRACE`) makes both the parser and the `*.bnf` diagrams substantially harder to write / comprehend.
> 
> I went for:
> 
> ```
> EXPLAIN OPTIMIZED QUERY GRAPH WITH (...) AS JSON ...
> EXPLAIN OPTIMIZED PLAN AS TEXT ...
> EXPLAIN OPTIMIZER TRACE WITH (...) AS TEXT ...
> ```
> 
> because @mjibson believes that `<thing to be explained>` should be first after the `EXPLAIN` verb.

I like that!

File: src/sql-parser/src/ast/defs/statement.rs [hunk 5]
```diff
@@ -2109,6 +2157,23 @@ impl<T: AstInfo> AstDisplay for Explainee<T> {
 }
 impl_display_t!(Explainee);
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum ExplainFormat {
```
Comment (aalexandrov) [inline src/sql-parser/src/ast/defs/statement.rs:2161]: nit: once we merge #13296 `ExplainFormat` and `ExplainConfig` will move into `mz-repr` along with the new `Explain` trait. 

We will need some glue code to instantiate `ExplainConfig` from a vector of parsed identifiers, but I can do that as part of the #13296 PR.

File: src/sql-parser/src/parser.rs [hunk 6]
```diff
@@ -4837,7 +4837,83 @@ impl<'a> Parser<'a> {
     /// Parse an `EXPLAIN` statement, assuming that the `EXPLAIN` token
     /// has already been consumed.
     fn parse_explain_new(&mut self) -> Result<Statement<Raw>, ParserError> {
-        Err(ParserError::new(self.index, "unimplemented"))
+        let stage = match self.parse_one_of_keywords(&[
```
Comment (aalexandrov) [inline src/sql-parser/src/parser.rs:4840]: Thanks for doing the extra pass, this is much easier to follow!

**Pull request merged**